crates/net/src/eventsource/futures.rs

```diff
@@ -0,0 +1,258 @@
+        match self.closures.lock() {
+            Ok(mut closures) => {
+                let (message_callbacks, _) = closures.deref_mut();
+
+                if let Some(message_callback) = message_callbacks.remove(event_type) {
+                    self.es
+                        .remove_event_listener_with_callback(
+                            event_type,
+                            message_callback.as_ref().unchecked_ref(),
+                        )
+                        .map_err(js_to_js_error)?;
+                }
+                Ok(())
+            }
+            Err(e) => Err(js_sys::Error::new(&format!("Failed to subscribe: {}", e)).into()),
+        }
+    }
+
+    /// Closes the EventSource.
+    ///
+    /// See the [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/close#parameters)
+    /// to learn about this function
+    pub fn close(self) {
+        self.es.close();
+    }
+
+    /// The current state of the EventSource.
+    pub fn state(&self) -> State {
+        let ready_state = self.es.ready_state();
+        match ready_state {
+            0 => State::Connecting,
+            1 => State::Open,
+            2 => State::Closed,
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+enum StreamMessage {
+    ErrorEvent,
+    Message(String, MessageEvent),
+}
+
+impl Stream for EventSource {
+    type Item = Result<(String, MessageEvent), EventSourceError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let msg = ready!(self.project().message_receiver.poll_next(cx));
+        match msg {
+            Some(StreamMessage::Message(event_type, msg)) => {
+                Poll::Ready(Some(Ok((event_type, msg))))
+            }
+            Some(StreamMessage::ErrorEvent) => {
+                Poll::Ready(Some(Err(EventSourceError::ConnectionError)))
+            }
+            None => Poll::Ready(None),
+        }
+    }
+}
+
+#[pinned_drop]
+impl PinnedDrop for EventSource {
+    fn drop(self: Pin<&mut Self>) {
+        self.es.close();
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::StreamExt;
+    use wasm_bindgen_futures::spawn_local;
+    use wasm_bindgen_test::*;
+
+    wasm_bindgen_test_configure!(run_in_browser);
+
+    const SSE_ECHO_SERVER_URL: &str = env!("SSE_ECHO_SERVER_URL");
+
+    #[wasm_bindgen_test]
+    fn eventsource_works() {
+        let mut es = EventSource::new(SSE_ECHO_SERVER_URL).unwrap();
+        es.subscribe_event("server").unwrap();
+        es.subscribe_event("request").unwrap();
+
+        spawn_local(async move {
+            assert_eq!(es.next().await.unwrap().unwrap().0, "server");
+            assert_eq!(es.next().await.unwrap().unwrap().0, "request");
+            es.unsubscribe_event("request").unwrap();
+        });
+    }
+
+    #[wasm_bindgen_test]
+    fn eventsource_close_works() {
```

Comment:
```suggestion
    fn eventsource_connection_failure_works() {
```

Testing close will mean `EventSource::close` is has test. Can you add a test for that?

Reply:
This test was named incorrectly. I've now renamed it to `eventsource_connect_failure_works`.

Commit 4688b89ac79c24124fef4a9460c9761509792100 addresses this.